File: api-gateway/gateway/src/main/java/com/example/gateway/SecurityConfig.java
package com.example.gateway;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.annotation.Order;
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.session.FlushMode;
import org.springframework.session.data.redis.RedisFlushMode;
import org.springframework.session.data.redis.config.annotation.web.http.EnableRedisHttpSession;

@Configuration
@EnableWebSecurity
@Order(1)
@EnableRedisHttpSession(
        flushMode = FlushMode.IMMEDIATE)
public class SecurityConfig extends WebSecurityConfigurerAdapter {
    @Autowired
    public void configureGlobal(AuthenticationManagerBuilder auth) throws Exception {
        System.out.println("In configureGlobal");
        auth.inMemoryAuthentication().withUser("user").password("{noop}password")
                .roles("USER").and().withUser("admin").password("{noop}admin")
                .roles("ADMIN").and().withUser("other").password("{noop}other").roles("OTHER");
        System.out.println("Out of configureGlobal");
    }

    @Override
    protected void configure(HttpSecurity http) throws Exception {
        System.out.println("In configure");
//        http.authorizeRequests().antMatchers("/movie-aggregator-service-rt")
//                .permitAll().antMatchers("/eureka/**").hasRole("ADMIN")
//                .anyRequest().authenticated().and().formLogin().and()
//                .logout().permitAll().logoutSuccessUrl("/movie-aggregator-service-rt")
//                .permitAll().and().csrf().disable();
        http.authorizeRequests().antMatchers("/movie-aggregator-service-rt/**")
                .hasAnyRole("USER", "ADMIN").antMatchers("/eureka/**").hasRole("ADMIN")
                .anyRequest().authenticated().and().formLogin().and()
                .logout().permitAll().logoutSuccessUrl("/movie-aggregator-service-rt")
                .permitAll().and().csrf().disable();
        http.sessionManagement().sessionCreationPolicy(SessionCreationPolicy.ALWAYS);
        System.out.println("Out of configure");
    }
}

File: microservice-workshop-kotlin/solution/eureka-server/src/main/kotlin/microservice/workshop/eurekaserver/SecurityConfig.kt
package microservice.workshop.eurekaserver

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.core.annotation.Order
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.AuthenticationEntryPoint
import org.springframework.security.web.authentication.www.BasicAuthenticationEntryPoint
import org.springframework.stereotype.Component
import java.io.IOException
import java.io.PrintWriter
import javax.servlet.ServletException
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse


@Configuration
//@EnableWebSecurity
@Order(1)
class SecurityConfig : WebSecurityConfigurerAdapter() {
    @Autowired
    fun configureGlobal(auth: AuthenticationManagerBuilder) {
        auth.inMemoryAuthentication().withUser("discUser")
            .password("{noop}discPassword").roles("SYSTEM")
        auth.inMemoryAuthentication().withUser("adminE").password("{noop}adminE").roles("ADMIN")
    }

    @Autowired
    val authenticationEntryPoint: MyBasicAuthenticationEntryPoint? = null

    override fun configure(http: HttpSecurity) {
        http.sessionManagement()
            .sessionCreationPolicy(SessionCreationPolicy.ALWAYS)
            .and().requestMatchers().antMatchers("/")
            .and().authorizeRequests().antMatchers("/")
            .hasRole("ADMIN").and().requestMatchers().antMatchers("/**")
            .and().authorizeRequests().antMatchers("/**").hasAnyRole("SYSTEM", "ADMIN")
            .anyRequest().denyAll().and()
            .httpBasic().authenticationEntryPoint(authenticationEntryPoint).and().csrf().disable()
        println("In SecurityConfig")
    }
}

@Component
class MyBasicAuthenticationEntryPoint : BasicAuthenticationEntryPoint() {

    override fun commence(
        request: HttpServletRequest, response: HttpServletResponse, authEx: AuthenticationException) {
        response.addHeader("WWW-Authenticate", "Basic realm=$realmName")
        response.status = HttpServletResponse.SC_UNAUTHORIZED
        val writer: PrintWriter = response.writer
        writer.println("HTTP Status 401 - " + authEx.message)
        println("This is in commence")
        println(request.getHeader("origin"))
        println(authEx.message)
    }

    override fun afterPropertiesSet(){
        realmName = "Test Realm Name";
        super.afterPropertiesSet();
    }
}

////@Configuration
//class AdminSecurityConfig :
//    WebSecurityConfigurerAdapter() {
//    override fun configure(http: HttpSecurity) {
//        http.sessionManagement()
//            .sessionCreationPolicy(SessionCreationPolicy.NEVER)
//            .and().httpBasic().disable().authorizeRequests()
//            .antMatchers(HttpMethod.GET, "/").hasRole("ADMIN")
//            .antMatchers("/info", "/health").authenticated().anyRequest()
//            .denyAll().and().csrf().disable()
//    }
//}
//@Configuration
//class SecurityConfig : WebSecurityConfigurerAdapter() {
//    override fun configure(http: HttpSecurity) {
//        http.authorizeRequests()
//            .anyRequest().permitAll()
//            .and()
//            .httpBasic().disable()
//            .csrf().disable()
//    }
//}
